File: csharp/Tennis/TennisGame4.cs
namespace Tennis;

public class TennisGame4  : ITennisGame
{
    internal int ServerScore;
    internal int ReceiverScore;
    internal readonly string Server;
    internal readonly string Receiver;

    public TennisGame4(string player1Name, string player2Name)
    {
        Server = player1Name;
        Receiver = player2Name;
    }

    public void WonPoint(string playerName)
    {
        if (Server.Equals(playerName))
            ServerScore += 1;
        else
            ReceiverScore += 1;
    }

    public string GetScore()
    {
        TennisResult result = new Deuce(
            this, new GameServer(
                this, new GameReceiver(
                    this, new AdvantageServer(
                        this, new AdvantageReceiver(
                            this, new DefaultResult(this)))))).GetResult();
        return result.Format();
    }

    internal bool ReceiverHasAdvantage() {
        return ReceiverScore >= 4 && (ReceiverScore - ServerScore) == 1;
    }

    internal bool ServerHasAdvantage() {
        return ServerScore >= 4 && (ServerScore - ReceiverScore) == 1;
    }

    internal bool ReceiverHasWon() {
        return ReceiverScore >= 4 && (ReceiverScore - ServerScore) >= 2;
    }

    internal bool ServerHasWon() {
        return ServerScore >= 4 && (ServerScore - ReceiverScore) >= 2;
    }

    internal bool IsDeuce() {
        return ServerScore >= 3 && ReceiverScore >= 3 && (ServerScore == ReceiverScore);
    }
}

internal class TennisResult {
    readonly string _serverScore;
    readonly string _receiverScore;

    public TennisResult(string serverScore, string receiverScore) {
        _serverScore = serverScore;
        _receiverScore = receiverScore;
    }

    internal string Format() {
        if ("".Equals(_receiverScore))
            return _serverScore;
        if (_serverScore.Equals(_receiverScore))
            return _serverScore + "-All";
        return _serverScore + "-" + _receiverScore;
    }
}

internal interface IResultProvider {
    TennisResult GetResult();
}

internal class Deuce : IResultProvider {
    private readonly TennisGame4 _game;
    private readonly IResultProvider _nextResult;

    public Deuce(TennisGame4 game, IResultProvider nextResult) {
        _game = game;
        _nextResult = nextResult;
    }

    public TennisResult GetResult() {
        if (_game.IsDeuce())
            return new TennisResult("Deuce", "");
        return _nextResult.GetResult();
    }
}

internal class GameServer : IResultProvider {
    private readonly TennisGame4 _game;
    private readonly IResultProvider _nextResult;

    public GameServer(TennisGame4 game, IResultProvider nextResult) {
        _game = game;
        _nextResult = nextResult;
    }

    public TennisResult GetResult() {
        if (_game.ServerHasWon())
            return new TennisResult("Win for " + _game.Server, "");
        return _nextResult.GetResult();
    }
}

internal class GameReceiver : IResultProvider {
    private readonly TennisGame4 _game;
    private readonly IResultProvider _nextResult;

    public GameReceiver(TennisGame4 game, IResultProvider nextResult) {
        _game = game;
        _nextResult = nextResult;
    }

    public TennisResult GetResult() {
        if (_game.ReceiverHasWon())
            return new TennisResult("Win for " + _game.Receiver, "");
        return _nextResult.GetResult();
    }
}

internal class AdvantageServer : IResultProvider {
    private readonly TennisGame4 _game;
    private readonly IResultProvider _nextResult;

    public AdvantageServer(TennisGame4 game, IResultProvider nextResult) {
        _game = game;
        _nextResult = nextResult;
    }

    public TennisResult GetResult() {
        if (_game.ServerHasAdvantage())
            return new TennisResult("Advantage " + _game.Server, "");
        return _nextResult.GetResult();
    }
}

internal class AdvantageReceiver : IResultProvider {

    private readonly TennisGame4 _game;
    private readonly IResultProvider _nextResult;

    public AdvantageReceiver(TennisGame4 game, IResultProvider nextResult) {
        _game = game;
        _nextResult = nextResult;
    }

    public TennisResult GetResult() {
        if (_game.ReceiverHasAdvantage())
            return new TennisResult("Advantage " + _game.Receiver, "");
        return _nextResult.GetResult();
    }
}

internal class DefaultResult : IResultProvider {

    private static readonly string[] Scores = {"Love", "Fifteen", "Thirty", "Forty"};

    private readonly TennisGame4 _game;

    public DefaultResult(TennisGame4 game) {
        _game = game;
    }

    public TennisResult GetResult() {
        return new TennisResult(Scores[_game.ServerScore], Scores[_game.ReceiverScore]);
    }
}

File: typescript-jest/src/TennisGame4.ts
import { TennisGame } from './TennisGame';

export class TennisGame4  implements TennisGame {

    public server : string;
    public receiver : string;
    public serverScore: number;
    public receiverScore: number;

    constructor(player1: string, player2: string) {
        this.server = player1;
        this.receiver = player2;
        this.serverScore = 0;
        this.receiverScore = 0;
    }

    getScore() {
        let result = new Deuce(
            this, new GameServer(
                this, new GameReceiver(
                    this, new AdvantageServer(
                        this, new AdvantageReceiver(
                            this, new DefaultResult(this)))))).getResult();
        return result.format();
    }

    receiverHasAdvantage() {
        return this.receiverScore >= 4 && (this.receiverScore - this.serverScore) === 1;
    }

    serverHasAdvantage() {
        return this.serverScore >= 4 && (this.serverScore - this.receiverScore) === 1;
    }

    receiverHasWon() {
        return this.receiverScore >= 4 && (this.receiverScore - this.serverScore) >= 2;
    }

    serverHasWon() {
        return this.serverScore >= 4 && (this.serverScore - this.receiverScore) >= 2;
    }

    isDeuce() {
        return this.serverScore >= 3 && this.receiverScore >= 3 && (this.serverScore === this.receiverScore);
    }

    wonPoint(playerName: string): void {
        if (playerName === 'player1')
            this.serverScore += 1;
        else
            this.receiverScore += 1;
    }
}

class TennisResult {
    private serverScore: string;
    private receiverScore: string;
    constructor(serverScore: string, receiverScore: string) {
        this.serverScore = serverScore;
        this.receiverScore = receiverScore;
    }

    format() {
        if ("" === this.receiverScore) {
            return this.serverScore;
        }
        if (this.serverScore === this.receiverScore) {
            return this.serverScore + "-All";
        }
        return this.serverScore + "-" + this.receiverScore;
    }
}

class Deuce {
    private game: TennisGame4;
    private nextResult: GameServer;
    constructor(game: TennisGame4, nextResult: GameServer) {
        this.game = game;
        this.nextResult = nextResult;
    }

    getResult() {
        if (this.game.isDeuce()) {
            return new TennisResult("Deuce", "");
        }
        return this.nextResult.getResult();
    }
}

class GameServer {
    private game: TennisGame4;
    private nextResult: GameReceiver;
    constructor(game: TennisGame4, nextResult: GameReceiver) {
        this.game = game;
        this.nextResult = nextResult;
    }

    getResult() {
        if (this.game.serverHasWon()) {
            return new TennisResult("Win for " + this.game.server, "");
        }
        return this.nextResult.getResult();
    }
}

class GameReceiver {
    private game: TennisGame4;
    private nextResult: AdvantageServer;
    constructor(game: TennisGame4, nextResult: AdvantageServer) {
        this.game = game;
        this.nextResult = nextResult;
    }

    getResult() {
        if (this.game.receiverHasWon()) {
            return new TennisResult("Win for " + this.game.receiver, "");
        }
        return this.nextResult.getResult();
    }
}

class AdvantageServer {
    private game: TennisGame4;
    private nextResult: AdvantageReceiver;
    constructor(game: TennisGame4, nextResult: AdvantageReceiver) {
        this.game = game;
        this.nextResult = nextResult;
    }

    getResult() {
        if (this.game.serverHasAdvantage()) {
            return new TennisResult("Advantage " + this.game.server, "");
        }
        return this.nextResult.getResult();
    }
}

class AdvantageReceiver {
    private game: TennisGame4;
    private nextResult: DefaultResult;
    constructor(game: TennisGame4, nextResult: DefaultResult) {
        this.game = game;
        this.nextResult = nextResult;
    }


    getResult() {
        if (this.game.receiverHasAdvantage()) {
            return new TennisResult("Advantage " + this.game.receiver, "");
        }
        return this.nextResult.getResult();
    }
}

class DefaultResult {
    private scores: string[];
    private game: TennisGame4;
    constructor(game: TennisGame4) {
        this.scores = ["Love", "Fifteen", "Thirty", "Forty"];
        this.game = game;
    }

    getResult() {
        return new TennisResult(this.scores[this.game.serverScore], this.scores[this.game.receiverScore]);
    }
}
